File: 2020/python2020/aoc/day02.py
#!/usr/bin/env python
"""
Advent Of Code 2020 Day 2
https://adventofcode.com/2020/day/2
"""
from typing import List, Optional, Tuple
import re


class Policy:
    def __init__(self, lower, upper, letter):
        self.lower = lower
        self.upper = upper
        self.letter = letter

    def __repr__(self):
        return f"<Policy {self.lower}, {self.upper}, {self.letter}>"

    def test1(self, password):
        return self.lower <= password.count(self.letter) <= self.upper

    def test2(self, password):
        cand_lower = password[self.lower - 1]
        cand_upper = password[self.upper - 1]
        return (cand_lower == self.letter) ^ (cand_upper == self.letter)


def parse(filename: str) -> List[Tuple[Policy, str]]:
    with open(filename) as file:
        return [parse_line(line.strip()) for line in file.readlines()]


def parse_line(line):
    left, password = line.split(": ")
    lower, upper, letter = re.search(r"(\d+)-(\d+) (\w)", left).groups()
    lower = int(lower)
    upper = int(upper)
    pol = Policy(lower, upper, letter)
    return (pol, password)


class Day02:
    """ AoC 2020 Day 02 """

    @staticmethod
    def part1(filename: str) -> int:
        """ Given a filename, solve 2020 day 02 part 1 """
        data = parse(filename)
        valid_pw_count = sum(1 for (pol, password) in data if pol.test1(password))
        return valid_pw_count

    @staticmethod
    def part2(filename: str) -> int:
        """ Given a filename, solve 2020 day 02 part 2 """
        data = parse(filename)
        valid_pw_count = sum(1 for (pol, password) in data if pol.test2(password))
        return valid_pw_count

File: 2020/elixir2020/lib/day02.ex
defmodule Elixir2020.Day02 do
  def parse(filename) do
    File.stream!(filename)
    |> Enum.map(&String.trim/1)
    |> Enum.map(&parse_rules/1)
  end

  def parse_rules(string) do
    [_, lower, upper, letter, password] = Regex.run(~r/^(\d+)-(\d+) (\w+): (\w+)$/, string)

    %{
      lower: String.to_integer(lower),
      upper: String.to_integer(upper),
      letter: letter,
      password: password
    }
  end

  def test1(%{lower: lower, upper: upper, letter: letter, password: password}) do
    count =
      password
      |> String.graphemes()
      |> Enum.filter(fn x -> x == letter end)
      |> Enum.count()

    lower <= count and count <= upper
  end

  def test2(%{lower: lower, upper: upper, letter: letter, password: password}) do
    lower_match = String.at(password, lower - 1) == letter
    upper_match = String.at(password, upper - 1) == letter
    lower_match != upper_match
  end

  def part1(filename) do
    filename
    |> parse()
    |> Enum.filter(&test1/1)
    |> Enum.count()
  end

  def part2(filename) do
    filename
    |> parse()
    |> Enum.filter(&test2/1)
    |> Enum.count()
  end
end
